### ISSUE TITLE
API to apply weights to gradients when using DistributedDataParallel

### ISSUE THREAD

**Issue by meder411** [pytorch/pytorch#31579]
## 🚀 Feature
I would like to propose that the API expose an option to weight the gradients during backpropagation in when using DistributedDataParallel.

## Motivation
Let’s say I have 4 GPUs and I am training a semantic segmentation network on a dataset with an ignore class. As I understand it, in the DataParallel setting, the outputs are aggregated on GPU0, the loss computed, and then the gradient is backpropagated back through each GPU’s model. In the DistributedDataParallel case, L0, L1, L2, L3 are each computed for each GPU’s share of the batch, the losses are backpropagated back through their respective GPU’s model, and the gradients are averaged along the way.

Using DataParallel, the presence of an ignore class makes no difference. Even if one GPU’s mini-batch has a lopsided amount of ignore pixels, the total loss is computed as the weighted average before any backprop occurs. However, what happens when you have a lopsided distribution of ignore pixels on one GPU using DistributedDataParallel? There does not seem to be any mechanism for weighting the average of the gradients; it seems that the averaging is unweighted. Yet in this case, L0, L1, L2, and L3 ought to have their contributions weighted by the ratio of valid pixels when averaging gradients during backpropagation.

## Pitch

I would like to see an optional field to `backward()` to provide weights for each backpropagation channel in DistributedDataParallel.

## Alternatives

Unclear what the alternatives are.

**Comment by mruberry** [comment]
Can you use the autograd hooks to weight the gradients before they're accumulated?

**Comment by meder411** [comment]
That’s a good alternative I hadn’t considered, although it’s quite manual for PyTorch.

**Comment by albanD** [comment]
@mruberry If I recall correctly, in case there are hooks. The same hook is called on each replica. So you might have to add the hooks manually on each replica to have different weights for each replica.

**Comment by mrshenli** [comment]
> Yet in this case, L0, L1, L2, and L3 ought to have their contributions weighted by the ratio of valid pixels when averaging gradients during backpropagation.

I might miss something here. If all you need is adding a weight to the DDP gradient average, will it be sufficient if you directly manipulate the loss? More specifically, instead of directly calling `Li.backward()`, can you do `(Li * wi).backward()`, where `wi` is the weight you would like to apply to gradients from GPUi?

Another thing I don't understand is why we need this weight in the first place. If you ignored some class/sample, I would expect those input data will not contribute to the local loss? If that is the case, the local gradients should already correctly representing that, and why would we need the weight here?

**Comment by pietern** [comment]
@meder411 On the one side there is the issue of how gradients are averaged across replicas when using `nn.DistributedDataParallel`. This is done with plain and simple averaging today without weighting of any kind. The averaging is an element-wise average of the gradients for every parameter in your model. For `nn.DataParallel`, there is no averaging to begin with. The different replicas simply sum to the same gradient tensor, so if a parameter isn't used in a particular replica, it won't contribute to the gradient.

On the other side you talk about ignore pixels, which leads me to think you're looking for a much more fine grained averaging than simply doing this for every parameter. Is this correct, or do you have data dependent control flow in your model causing some subset of parameters to optionally go unused?

If it is the latter, i.e. you have unused parameters, we could figure out a way to make the averaging dependent on the number of replicas that contributed to the gradient. This would change the current behavior from scaling all gradients by `1 / N` to scaling every gradient individually by `1 / N'` where `N'` is the number of replicas that contributed to a particular gradient.

Can you confirm?

**Comment by meder411** [comment]
Sorry for the months-long delay on this.

@pietern I'm not doing anything particularly technical. It's actually just a matter of simple semantic segmentation with a small u-net network. I'm using identical code, except one is `DistributedDataParallel` while the other is old-fashioned `DataParallel`. No matter what I do, I cannot get equal performance. See the image below:

![image](https://user-images.githubusercontent.com/6818607/80662370-1a8a0300-8a5f-11ea-8c66-86a0ff0fd8c1.png)

The lines prefixed with "0-..." are with `DataParallel` and the lines prefixed with "1-..." are `DistributedDataParallel`. I've run both numerous times. Each variant confirms itself, but they never agree with each other. Talking it over and inpsecting the code with some of my colleagues, all we could come up with was that it might be something to do with the way the gradients are merged Even from the first epoch, validation loss is significantly lower with distributed than with `DataParallel`. It's only a 13 class dataset, but I'd say about 15-20% of the pixels are ignored in any given image. We have seen this behavior on variant datasets and networks too.  Given a trained model, the eval results are identical using both `DataParallel` and `DistributedDataParallel`

If it's not an issue with gradients, then I'm completely at a loss. Of course it's always possible I'm doing something wrong with distributed somewhere else in training, but I haven't been able to identify anything, which is why I proposed this issue.

**Comment by meder411** [comment]
Sorry for the months-long delay on this.

@pietern I'm not doing anything particularly technical. It's actually just a matter of simple semantic segmentation with a small u-net network. I'm using identical code, except one is DistributedDataParallel while the other is old-fashioned DataParallel. No matter what I do, I cannot get equal performance. See the image below:

![image](https://user-images.githubusercontent.com/6818607/80663267-81102080-8a61-11ea-9733-bb79f3b37cef.png)

The lines prefixed with "0-..." are with DataParallel and the lines prefixed with "1-..." are DistributedDataParallel. I've run both numerous times. Each variant confirms itself, but they never agree with each other. Talking it over and inspecting the code with some of my colleagues, all we could come up with was that it might be something to do with the way the gradients are merged. Even from the first epoch, validation loss is significantly lower with distributed than with DataParallel, even though training loss is near-identical. It's only a 13 class dataset, but I'd say about 15-20% of the pixels are ignored in any given image. We have seen this behavior on variant datasets and networks too. Given a trained model, the eval results are identical using both DataParallel and DistributedDataParallel. 

If it's not an issue with gradients, then I'm completely at a loss. Of course it's always possible I'm doing something wrong with distributed somewhere else in training, but I haven't been able to identify anything, which is why I proposed this issue.

**Comment by meder411** [comment]
Ah, I actually just came across [this thread](https://discuss.pytorch.org/t/training-performance-degrades-with-distributeddataparallel/47152/4), which @pietern  actually commented on last summer. It seems I'm not alone in this. I will try the LR / world_size or synced BN issue. Perhaps I was barking up the wrong tree with the gradient weighting.
